pydicom/encoders/base.py

```diff
@@ -0,0 +1,779 @@
+"""Bulk data encoding."""
+
+from importlib import import_module
+import sys
+from typing import (
+    Callable, Generator, Tuple, List, Optional, Dict, Union, cast
+)
+
+from pydicom.dataset import Dataset
+from pydicom.encaps import encapsulate
+from pydicom.pixel_data_handlers.util import get_expected_length
+from pydicom.uid import UID, RLELossless
+
+try:
+    import numpy
+    import numpy as np
+except ImportError:
+    pass
+
+
+class Encoder:
+    """Factory class for data encoders.
+
+    Every available ``Encoder`` instance in *pydicom* corresponds directly
+    to a single DICOM *Transfer Syntax UID*, and provides a
+    mechanism for converting raw unencoded source data to meet the
+    requirements of that transfer syntax using one or more
+    :doc:`encoding plugins</guides/encoder_plugins>`.
+
+    .. versionadded:: 2.2
+    """
+    def __init__(self, uid: UID) -> None:
+        """Create a new data encoder.
+
+        Parameters
+        ----------
+        uid : pydicom.uid.UID
+            The *Transfer Syntax UID* that the encoder supports.
+        """
+        # The *Transfer Syntax UID* data will be encoded to
+        self._uid = uid
+        # Available encoding plugins
+        self._available: Dict[str, Callable] = {}
+        # Unavailable encoding plugins - missing dependencies or other reason
+        self._unavailable: Dict[str, Tuple[str, ...]] = {}
+        # Default encoding options
+        self._defaults = {
+            'transfer_syntax_uid': self.UID,  # Intended transfer syntax
+            'byteorder': '<',  # Byte ordering of `src` passed to plugins
+        }
+
+    def add_plugin(self, label: str, import_path: Tuple[str, str]) -> None:
+        """Add an encoding plugin to the encoder.
+
+        The requirements for encoding plugins are available
+        :doc:`here</guides/encoder_plugins>`.
+
+        Parameters
+        ----------
+        label : str
+            The label to use for the plugin, should be unique for the encoder.
+        import_path : Tuple[str, str]
+            The module import path and the encoding function's name (e.g.
+            ``('pydicom.encoders.pylibjpeg', 'encode_pixel_data')``).
+        """
```

Comment:
Maybe add that it will raise `ModuleNotFoundError` for an invalid `import_path`.

Reply:
Done